### DISCUSSION TITLE
(Questions) Add zoomIn and zoomOut funcionality on external button and block zoom if X

### DISCUSSION THREAD

**Discussion by Salsao** [samuelarbibe/dnd-timeline#46] [Q&A] [answered]
Hi there!

I'm working on adding zoom in and zoom out buttons to adjust the zoom/scale of the timeline, but I haven't figured it out yet. Can you help me with this?

Also, since I'm dealing with a video timeline, I don't want the values to go below 0 (like starting at 0 minutes and ending at 5 minutes). When I zoom out with the mouse wheel, it goes infinitely. Is there a way to prevent this?

I'm using <TimelineContext>

Thanks!

**Comment by samuelarbibe** [comment, marked as answer]
Hey there!  
The timeline is fully controlled by an external state, managed by you, and the viewed area can be changed by changing the provided `range` prop.   
You should keep in mind that "zoom" is actually the range size:  The bigger the range, the smaller the zoom.  
If you want to "zoom in", you need to make the range smaller. You can do that by subtracting the range on each side by some amount.  
Similarly, "zooming out" consists of enlarging the range provided to the timeline context.   
  
As for limiting the zooming, you can do so by actually updating the range state **only** if you predefined limitations apply to it.  
For example, If you want to limit the range to always be positive, you can do like so:  

```tsx
  const [range, setRange] = useState(DEFAULT_RANGE)
 
  const handleOnRangeChanged = (setterFunc) => {
    setRange((prev) => {
      const next = setterFunc(prev);
      return {
        start: Math.max(0, next.start),
        next: next.end
      }
    })
  }

  return (
    <TimelineContext
      onResizeEnd={onResizeEnd}
      range={range}
      onRangeChanged={handleOnRangeChanged}
    >
      <Timeline rows={rows} items={items} />
    </TimelineContext>
  )
```
